## Running Performance Optimization

### 1. Data Structure 
### 2. Determistic -> Heuristic & Approximation Algorithm 
### 3. Simulation to solve too complicated problems. 
### 3. Cache 


Case 1.  Find an element if exists in a list. 

In [7]:
def find_exist(element, L):
    for e in L:
        if e == element:
            return e
    return False

## Time complexity: O(n)

In [2]:
import random

In [10]:
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __int__(self, /)
 |      int(sel

In [16]:
large_list = [round(random.uniform(-1000, 1000), 1) for _ in range(100000)]

In [27]:
import time

In [33]:
sorted(large_list)

[-1000.0,
 -1000.0,
 -1000.0,
 -1000.0,
 -1000.0,
 -999.9,
 -999.9,
 -999.9,
 -999.9,
 -999.9,
 -999.9,
 -999.9,
 -999.8,
 -999.8,
 -999.7,
 -999.7,
 -999.7,
 -999.7,
 -999.5,
 -999.5,
 -999.5,
 -999.5,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.4,
 -999.3,
 -999.2,
 -999.2,
 -999.2,
 -999.2,
 -999.2,
 -999.1,
 -999.1,
 -999.1,
 -999.1,
 -999.1,
 -999.1,
 -999.0,
 -999.0,
 -999.0,
 -999.0,
 -999.0,
 -998.9,
 -998.9,
 -998.9,
 -998.8,
 -998.8,
 -998.8,
 -998.8,
 -998.8,
 -998.7,
 -998.7,
 -998.7,
 -998.6,
 -998.6,
 -998.6,
 -998.6,
 -998.6,
 -998.6,
 -998.6,
 -998.5,
 -998.5,
 -998.5,
 -998.5,
 -998.5,
 -998.5,
 -998.5,
 -998.5,
 -998.4,
 -998.3,
 -998.3,
 -998.3,
 -998.2,
 -998.2,
 -998.2,
 -998.2,
 -998.2,
 -998.1,
 -998.1,
 -998.1,
 -998.0,
 -998.0,
 -998.0,
 -998.0,
 -998.0,
 -998.0,
 -998.0,
 -998.0,
 -997.9,
 -997.9,
 -997.9,
 -997.9,
 -997.8,
 -997.8,
 -997.8,
 -997.8,
 -997.8,
 -997.7,
 -997.7,
 -997.7,
 -997.7,
 -997.6,
 -997.6,
 -997.5,
 -997

In [ ]:
def distinctive(elements):
    sorted_elements = sorted(elements)
    ## 

In [29]:
begin = time.time()
for i in range(1000):
    n = round(random.uniform(-1000, 1000), 2)
    find_exist(n, large_list)

print(f'time consumed: {time.time() - begin}')

time consumed: 2.148271083831787


In [45]:
def test_in_elements(e, L):
    return e in L

In [58]:
large_ints = [random.randint(-10000, 10000) for _ in range(100000)]

In [64]:
array_ints = array('h', large_ints)

In [65]:
set_ints = set(large_ints)

In [70]:
begin = time.time()
for i in range(1000):
    n = random.randint(-100000, 100000)
    test_in_elements(n, set_ints)

print(f'time consumed: {time.time() - begin}')

time consumed: 0.006934165954589844


| strucute-type | dta-type | time | 
| -- | -- | -- | 
| list |  float | 0.84 |
| set | float | 0.01 | 
| array | float | 1.74 | 
| list |  int | 1.23 |
| set | int | .0069 | 
| array | int | 2.62 | 

In [36]:
from array import array

In [57]:
help(array)

Help on class array in module array:

class array(builtins.object)
 |  array(typecode [, initializer]) -> array
 |  
 |  Return a new array whose items are restricted by typecode, and
 |  initialized from the optional initializer value, which must be a list,
 |  string or iterable over elements of the appropriate type.
 |  
 |  Arrays represent basic values and behave very much like lists, except
 |  the type of objects stored in them is constrained. The type is specified
 |  at object creation time by using a type code, which is a single character.
 |  The following type codes are defined:
 |  
 |      Type code   C Type             Minimum size in bytes 
 |      'b'         signed integer     1 
 |      'B'         unsigned integer   1 
 |      'u'         Unicode character  2 (see note) 
 |      'h'         signed integer     2 
 |      'H'         unsigned integer   2 
 |      'i'         signed integer     2 
 |      'I'         unsigned integer   2 
 |      'l'         signed int

In [42]:
large_array = array('f', large_list)

In [56]:
dir(large_array)

['__add__',
 '__class__',
 '__contains__',
 '__copy__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'buffer_info',
 'byteswap',
 'count',
 'extend',
 'frombytes',
 'fromfile',
 'fromlist',
 'fromstring',
 'fromunicode',
 'index',
 'insert',
 'itemsize',
 'pop',
 'remove',
 'reverse',
 'tobytes',
 'tofile',
 'tolist',
 'tostring',
 'tounicode',
 'typecode']

In [34]:
[1, 2, 'test', 'some other', [1, 2, 3, 4]]

[1, 2, 'test', 'some other', [1, 2, 3, 4]]

## deque, double ending queue

In [91]:
total_length = 1000
message_queue = [None] * total_length

In [92]:
from collections import deque

In [93]:
deque_msg = deque(message_queue)

In [96]:
def msg_control(running=1000):
    for i in range(running):
        new_msg_num = random.randint(0, 10)
        while new_msg_num:
            #message_queue.pop(0)
            deque_msg.popleft()
            t = time.time()
            #message_queue.append(t)
            deque_msg.append(t)
            new_msg_num -= 1

In [97]:
%timeit msg_control()

1.82 ms ± 26.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


| structure-type | length | time |
| -- | --  | -- | 
| list | 1000 | 2.85 ms | 
| deque | 1000 | 1.82 ms | 
| list | 1e6 | 123ms |
| deque | 1e6 | 1.78ms | 


## defaultdict 

In [104]:
some_dict = dict()

In [111]:
some_dict[1] = 100

In [112]:
some_dict.setdefault(1, 0) # => if 1 in some_dict: 

100

In [114]:
if 1 in some_dict: 
    pass
else:
    some_dict[1] = 0

In [115]:
from collections import defaultdict

In [137]:
def _100(): return 100

In [138]:
some_defalutdict = defaultdict(_100)

In [120]:
%timeit some_dict.setdefault(10, 0)

64.5 ns ± 0.777 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [122]:
%timeit some_defalutdict[10]

33.6 ns ± 0.413 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [123]:
import pickle

In [124]:
help(pickle)

Help on module pickle:

NAME
    pickle - Create portable serialized representations of Python objects.

MODULE REFERENCE
    https://docs.python.org/3.7/library/pickle
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    See module copyreg for a mechanism for registering custom picklers.
    See module pickletools source for extensive comments.
    
    Classes:
    
        Pickler
        Unpickler
    
    Functions:
    
        dump(object, file)
        dumps(object) -> string
        load(file) -> object
        loads(string) -> object
    
    Misc variables:
    
        __version__
        format_version
        compatible_formats

CLASSES
    builtins.Exception(builtins.BaseException)
     

In [139]:
with open('test-pickle.pkl', 'wb') as f:
    pickle.dump(some_defalutdict, f)

In [141]:
with open('test-pickle.pkl', 'rb') as f:
    print(pickle.load(f))

defaultdict(<function _100 at 0x7fc718998d90>, {})


## nametuple

In [143]:
class Customer: 
    def __init__(self, name, age, rank):
        self.name = name
        self.age = age
        self.rank = rank

In [193]:
from dataclasses import dataclass

In [194]:
@dataclass
class DCustomer: 
    name : str
    age : int
    rank : int

In [171]:
from string import ascii_letters

In [172]:
def first_upper(string):
    return string[0].upper() + string[1:].lower()

In [173]:
def gene_name(): 
    length = random.randint(3, 6)
    return first_upper(''.join(random.sample(ascii_letters, length)))

In [188]:
gene_name()

'Htlriz'

In [196]:
from collections import namedtuple

In [198]:
TCustomer = namedtuple(
    'Customer', 
    'name age rank'
)

In [210]:
%%timeit 

customers = []

for _ in range(1000):
    TCustomer(
        name=gene_name(),
        age=random.randint(5, 80),
        rank=random.randint(0, 10)
    )


#ranks = []

#for i, c in enumerate(customers):
#    ranks.append(c.rank)

7.32 ms ± 45.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


| type | length | time | 
|-- | -- | -- | 
|class | 1000| 7.34 ms  |  
|Dclass | 1000 | 7.28 ms ms  | 
| namedtuple | 1000 | 7.26 ms | 

In [211]:
string = 'somewords'

## 近似化（往往伴随了随机化）

有很多复杂问题，如果要获得其精确解，那么需要的求解是非常复杂的。 如果我们要获得近似解，则，难度会降低很多~

## 数值分析(numerical analysis) 是进行算法优化非常重要的方式

### 2018有一个银行，要做运钞车的线路优化问题

In [214]:
def fac(n):
    return n * fac(n - 1) if n > 1 else 1

In [224]:
fac(200) 

788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000

In [225]:
5 * fac(40)

4079576416239488671728056347980579471360000000000

## Cache 

## 1. functools.cache 
## 2. Web: file system, db
## 3. https://github.com/pinterest/pymemcache